### PR TITLE
Dist version

### LINKED ISSUE #7: Dist version specific graphs
It would be nice to be able to view dep graphs for different releases of a distribution rather than only keeping the latest release of each distribution.

### PR DESCRIPTION
**Pull request by mohawk2**
I believe this will fix #7

### CODE REVIEW

File: cpan_deps_graph.pl [hunk 1]
```diff
@@ -96,16 +105,21 @@
...
   my $redis = $c->redis->db;
   my %all_deps;
   foreach my $phase (@$phases) {
     foreach my $relationship (@$relationships) {
-      my $key = "cpandeps:$dist:$phase:$relationship";
-      my $deps_json = $redis->get($key) // next;
       my $deps;
-      try { $deps = from_json $deps_json } catch { next }
+      if ($dist_version) {
+        $deps = $c->retrieve_dist_deps($dist, $dist_version);
```
Comment (Grinnz) [inline cpan_deps_graph.pl]: This should only need to be done once rather than for each phase/relationship combo

File: cpan_deps_graph.pl [hunk 2]
```diff
@@ -117,32 +131,33 @@
   return \%all_deps;
 };
 
-helper dist_dep_tree => sub ($c, $dist, $phases, $relationships, $perl_version) {
+helper dist_dep_tree => sub ($c, $dist, $phases, $relationships, $perl_version, $dist_version = undef) {
   my %seen;
   my %deps;
-  my @to_check = $dist;
-  while (defined(my $dist = shift @to_check)) {
+  my @to_check = [ $dist, $dist_version ]; # only pass on version for initial
```
Comment (Grinnz) [inline cpan_deps_graph.pl]: I'd prefer this meta structure was a hashref than an arrayref for clarity

File: cpan_deps_graph.pl [hunk 3]
```diff
@@ -96,16 +105,21 @@
...
   $perl_version = $perl_version->numify;
   my $redis = $c->redis->db;
   my %all_deps;
   foreach my $phase (@$phases) {
     foreach my $relationship (@$relationships) {
-      my $key = "cpandeps:$dist:$phase:$relationship";
-      my $deps_json = $redis->get($key) // next;
       my $deps;
-      try { $deps = from_json $deps_json } catch { next }
+      if ($dist_version) {
```
Comment (Grinnz) [inline cpan_deps_graph.pl]: This needs to be 0-safe - at the very least, check length $dist_version if not defined

File: public/graph.js [hunk 4]
```diff
@@ -85,6 +86,7 @@ function retrieve_graph() {
 
   var deps_url = new URL('/api/v1/deps', window.location.href);
   deps_url.searchParams.set('dist', dist);
+  deps_url.searchParams.set('dist_version', dist_version);
```
Comment (Grinnz) [inline public/graph.js]: Would be better to only set this if dist_version was not null or empty string

File: cpan_deps_graph.pl (hunk 3 above)
Comment (mohawk2) [reply cpan_deps_graph.pl]: Assuming you mean "if defined" - agreed, will execute.

File: cpan_deps_graph.pl (hunk 1 above)
Comment (mohawk2) [reply cpan_deps_graph.pl]: Done

File: public/graph.js (hunk 4 above)
Comment (mohawk2) [reply public/graph.js]: Done

File: cpan_deps_graph.pl (hunk 2 above)
Comment (mohawk2) [reply cpan_deps_graph.pl]: Done

File: cpan_deps_graph.pl [hunk 5]
```diff
@@ -96,16 +104,22 @@
   }
 };
 
-helper get_dist_deps => sub ($c, $dist, $phases, $relationships, $perl_version) {
+helper get_dist_deps => sub ($c, $dist, $phases, $relationships, $perl_version, $dist_version = undef) {
   $perl_version = $perl_version->numify;
   my $redis = $c->redis->db;
   my %all_deps;
+  my $versioned_deps = $c->retrieve_dist_deps($dist, $dist_version)
```
Comment (Grinnz) [inline cpan_deps_graph.pl]: conditional 'my' declaration, and you can simplify the test to `length $dist_version` as this won't be running on perl 5.10

Comment (mohawk2) [reply cpan_deps_graph.pl]: I believe the new version fixes this? (My thinking was it's not a conditional declaration, but a conditional assignment, but it's up to you :-)

Comment (Grinnz) [reply cpan_deps_graph.pl]: any "my ... if ..." is problematic. LGTM

**Pull request merged**